<a href="https://colab.research.google.com/github/Ayubbett/Moringa-School-Week-9-Independent-Project/blob/main/Moringa_School_Data_Science_Core_W9_Independent_Project_2021_12_AYUB_BETT_Python_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **SPAM DETECTION**

**1. Defining the Question**

**a) Specifying the Data Analytic Question**

An email is categorised as being a spam or not depending on the the content it has. So in thi problem we are to categorise the classification of an email being spam or not depending on what it contains.

**b) Defining the Metric for Success**

We are going to build a Naive Bayes classifier to detect spam emails from the others. We are going to put the threshold for having a reliable Email at an accuracy of between 75% and 95%. Above that we are going to consider it overfitting the data and below it we shall conclude it underfitting.

**c) Understanding the context**

**d) Recording the Experimental Design**

Below are the steps we are going to take to approach our modelling:

*   Load the necessary libraries for visualisation and data manipulation.
*   Load the datset and view its columns and rows.

*   Clean the data by looking at the missing values, duplicated values and any anomalies within the data.
*   Perform exploratory data analysis that is Univariate analysis and Bivariate analysis.

*   Perform feature engineering to get the best features to use.
*   Perform modelling using Naive Bayes Classifier.







**e) Data Relevance**

**2. Reading the Data**

In [2]:
#Importing the required libraries
import pandas as pd
import numpy as np

#columns = [['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d', 'word_freq_our', 'word_freq_over', 'word_freq_remove',
#           'word_freq_internet', 'word_freq_order', 'word_freq_mail', 'word_freq_receive', 'word_freq_will', 'word_freq_people',
#           'word_freq_report', 'word_freq_addresses', 'word_freq_free', 'word_freq_business', 'word_freq_email', 'word_freq_you', 
#          'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000', 'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 
#           'word_freq_george', 'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet', 'word_freq_857', 'word_freq_data', 
#          'word_freq_415', 'word_freq_85', 'word_freq_technology', 'word_freq_1999', 'word_freq_parts', 'word_freq_pm', 'word_freq_direct',
#           'word_freq_cs', 'word_freq_meeting', 'word_freq_original', 'word_freq_project', 'word_freq_re', 'word_freq_edu', 'word_freq_table',
#           'word_freq_conference', 'char_freq_colon', 'char_freq_parenthesis', 'char_freq_square', 'char_freq_apost', 'char_freq_dollar', 'char_freq_hash', 
#         'capital_run_length_average', 'capital_run_length_longest', 'capital_run_length_total']]                                                         


df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/spambase/spambase.data')
df.head(1)

0  0.64  0.64.1  0.1  0.32   0.2  ...  0.43   0.44  3.756   61   278  1
0  0.21  0.28     0.5  0.0  0.14  0.28  ...  0.18  0.048  5.114  101  1028  1

[1 rows x 58 columns]

In [3]:
df.rename(columns={'0':'word_freq_make', '0.64':'word_freq_address', '0.64.1':'word_freq_all', '0.1':'word_freq_3d','0.32':'word_freq_our', '0.2':'word_freq_over', '0.3':'word_freq_remove',
           '0.4':'word_freq_internet', '0.5':'word_freq_order', '0.6':'word_freq_mail', '0.7':'word_freq_receive', '0.64.2':'word_freq_will', '0.8':'word_freq_people',
           '0.9':'word_freq_report', '0.10':'word_freq_addresses', '0.32.1':'word_freq_free', '0.11':'word_freq_business', '1.29':'word_freq_email', '1.93':'word_freq_you', 
          '0.12':'word_freq_credit', '0.96':'word_freq_your', '0.13':'word_freq_font', '0.14':'word_freq_000', '0.15':'word_freq_money', '0.16':'word_freq_hp', '0.17':'word_freq_hpl', 
           '0.18':'word_freq_george', '0.19':'word_freq_650', '0.20':'word_freq_lab', '0.21':'word_freq_labs', '0.22':'word_freq_telnet', '0.23':'word_freq_857', '0.24':'word_freq_data', 
          '0.25':'word_freq_415', '0.26':'word_freq_85', '0.27':'word_freq_technology', '0.28':'word_freq_1999', '0.29':'word_freq_parts', '0.30':'word_freq_pm', '0.31':'word_freq_direct',
           '0.32':'word_freq_cs', '0.33':'word_freq_meeting', '0.34':'word_freq_original', '0.35':'word_freq_project', '0.36':'word_freq_re', '0.37':'word_freq_edu', '0.38':'word_freq_table',
           '0.39':'word_freq_conference', '0.40':'char_freq_colon', '0.41':'char_freq_parenthesis', '0.42':'char_freq_square', '0.778':'char_freq_apost', '0.43':'char_freq_dollar', '0.44':'char_freq_hash', 
         '3.756':'capital_run_length_average', '61':'capital_run_length_longest', '278':'capital_run_length_total', '1':'Spam'}, inplace=True)

In [4]:
df.head()

word_freq_make  word_freq_address  ...  capital_run_length_total  Spam
0            0.21               0.28  ...                      1028     1
1            0.06               0.00  ...                      2259     1
2            0.00               0.00  ...                       191     1
3            0.00               0.00  ...                       191     1
4            0.00               0.00  ...                        54     1

[5 rows x 58 columns]

**3. Checking the Data**

In [5]:
df.shape

(4600, 58)

In [6]:
# Checking whether each column has an appropriate datatype
#
df.dtypes

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_cs                  float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 float64
word_freq_ge

**4. External Data Source Validation**

By checking other sources having the same dataset we notice that we have what can be considered the best version of the data we can get.

**5. Tidying the Dataset**

In [7]:
# Identifying the Missing Data
#
df.isnull().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_cs                  0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

We have no null values.

In [8]:
#Checking for duplicated values
df.duplicated().sum()

391

**6. Exploratory Analysis**

We notice that this dataset has been presented in  a way that it is diffucult to perform Exploratory Data Analysis, So we are going to do its analysis straight away.

**7. Implementing the Solution**

We notice that some entries are zero so we smoothe the dataset to avoid zero frequency error.

In [9]:
#Creating target columns with feature set
X = df.drop(['Spam'], axis=1)
y = df['Spam']

**7.1. Using 80:20 ratio**

In [10]:
#Splitting our data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0) 

In [11]:
#since all the column values are continous we are going to use Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()  
model = clf.fit(X_train, y_train)
y_pred1 = model.predict(X_test)

In [12]:
#Checking the accuracy of our model
from sklearn.metrics import classification_report, confusion_matrix

print(confusion_matrix(y_test, y_pred1))
print(classification_report(y_test, y_pred1))

[[392 146]
 [ 10 372]]
              precision    recall  f1-score   support

           0       0.98      0.73      0.83       538
           1       0.72      0.97      0.83       382

    accuracy                           0.83       920
   macro avg       0.85      0.85      0.83       920
weighted avg       0.87      0.83      0.83       920



Using a 80:20 ratio we find a accuracy of 83%. The precision and recall values seem to be varying quite much.

**7.2. Using 70:30 Ratio**

In [13]:
#Splitting our data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [16]:
#since all the column values are continous we are going to use Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()  
model = clf.fit(X_train, y_train)
y_pred2 = model.predict(X_test)

In [17]:
#Checking the accuracy of our model
from sklearn.metrics import classification_report, confusion_matrix

print(confusion_matrix(y_test, y_pred2))
print(classification_report(y_test, y_pred2))

[[600 222]
 [ 15 543]]
              precision    recall  f1-score   support

           0       0.98      0.73      0.84       822
           1       0.71      0.97      0.82       558

    accuracy                           0.83      1380
   macro avg       0.84      0.85      0.83      1380
weighted avg       0.87      0.83      0.83      1380



The accuracy score does not change from the previous ratio.

**7.3. Using 60:40 ratio**

In [18]:
#Splitting our data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)

In [19]:
#since all the column values are continous we are going to use Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()  
model = clf.fit(X_train, y_train)
y_pred3 = model.predict(X_test)

In [20]:
#Checking the accuracy of our model
from sklearn.metrics import classification_report, confusion_matrix

print(confusion_matrix(y_test, y_pred3))
print(classification_report(y_test, y_pred3))

[[791 306]
 [ 25 718]]
              precision    recall  f1-score   support

           0       0.97      0.72      0.83      1097
           1       0.70      0.97      0.81       743

    accuracy                           0.82      1840
   macro avg       0.84      0.84      0.82      1840
weighted avg       0.86      0.82      0.82      1840



We note that using a ratio of 60:40 reduces the accuracy slightly to 82%

**8. Conclusion**

We notice that changing the test and train size doesn't much affect the accuracy when working with naive bayes classifier.